src/main/java/InfrastructureManager/EdgeClientHistory.java

```diff
@@ -0,0 +1,163 @@
+package InfrastructureManager;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * This class is dedicated to fetch history related info.
+ * From get history score, to history time to calculate the score.
+ * Implemented function : get historyScore - get ConnectedTime
+ * Also set historyMap and set ConnectedMap
+ */
+public class EdgeClientHistory {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    // - -------------------------------- ---- THIS IS THE CONSTRUCTOR -----------------------------
+    public EdgeClientHistory(String id) {
+        this.id = id;
+    }
+
+    //put client ID in here, put client's connected node and node's score.
+    //need to get time where client's connected to node
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public Multimap<String, HashMap<String, Long>> getHistoryMap() {
+        return historyMap;
+    }
+
+    /**
+     * Assign client, node and history score into class's multimap.
+     * This function create a new HashMap, use mainly for initialize out collection.
+     * Don't use to replace value, because we are using multimap, 1 key can have several similar value
+     * @param clientID
+     * @param nodeID
+     * @param nodeHistory
+     */
+    public void setHistoryMap(String clientID, String nodeID, Long nodeHistory) {
+        HashMap<String, Long> nodeAndScore = new HashMap<>();
+        nodeAndScore.put(nodeID, nodeHistory);
+        this.historyMap.put(clientID, nodeAndScore);
+    }
+
+    public void setHistoryScoreForClient(String clientID, String nodeID, Long nodeHistoryScore){
+        Collection<HashMap<String, Long>> clientHistoryMapList = this.historyMap.get(clientID);
+        for (HashMap<String, Long> theList : clientHistoryMapList){
```

Comment:
Unfortunately, this is the most optimal way I can do tbh, 
There aren't any matrix structures (i guess you mean like SQL data table) that which native to Java 😆 and accessed by key. So either we stuck with the String[ ] [ ] 2x2 matrix like this, and iterate the same way to find the match (using .equals()) and extract the value.
And the iteration through the list is actually not that slow tbh even if we scale up (i once run a program iterate 1mil entry keys and only took 30s to finish the program, let alone the iteration only).
Also, the break at the end of the iteration will take us out of the loop as soon as we find a match, removing any unnecessary movement. 
I don't see any possibility to use other methods to access a target string faster than this tbh.

Reply:
Also even for matrix, when finding a key, it also has to iterate most of it's contained keys in the object, and check whether that match without input parameter or not. The process is the same. Since even for a String to be assigned to an address/pointer :)) we can't say 2 String is the same because they are in the same address, so iterate one by one is inevitable.
Unless someone find a way that is superior to the method above, then I think just leave it alone. @alyhasansakr

Reply:
Of course there is no faster solution as it is. Building an index is one of the ways you can make it faster (same as used in databases), this can reduce the time from n to log(n) for each dimension.

Anyway, just leave a TODO here.

Reply:
Done in 226b692 
will come back and test new table (guava tablemap / HashBasedTable)